File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Map<TreeNode,TreeNode> map=new HashMap<>();
        map.put(root,null);
        dfs(root,map);
        Set<TreeNode> set=new HashSet<>();
        while(p!=null){
            set.add(p);
            p=map.get(p);
        }
        while(q!=null){
            if(set.contains(q))return q;
            set.add(q);
            q=map.get(q);
        }
        return null;
    }
    void dfs(TreeNode curr,Map<TreeNode,TreeNode> map){
        if(curr.left!=null){
            map.put(curr.left,curr);
            dfs(curr.left,map);
        }
        if(curr.right!=null){
            map.put(curr.right,curr);
            dfs(curr.right,map);
        }
    }
}

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        
        # Using Recursion
        if (root == p or root == q or not root): return root
        
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p , q)
        
        if left and right : return root  #Found common ancestor
        
        elif left: return left
        elif right: return right
